newlib/libc/include/sys/cdefs.h

```diff
@@ -558,11 +558,53 @@
 #define	__strong_reference(sym,aliassym)	\
 	extern __typeof (sym) aliassym __attribute__ ((__alias__ (#sym)))
 #endif
+
 #ifdef __ELF__
 #ifdef __STDC__
 #define	__weak_reference(sym,alias)	\
 	__asm__(".weak " #alias);	\
 	__asm__(".equ "  #alias ", " #sym)
+#else
+#define	__weak_reference(sym,alias)	\
+	__asm__(".weak alias");		\
+	__asm__(".equ alias, sym")
+#endif
+#elif __clang__
+#ifdef __MACH__
+/* Macos prefixes all C symbols with an underscore, this needs to be done manually in asm */
+
+/* So far I have not been able to create on Macos an exported symbol that itself
+ * is weak but aliases a strong symbol. A workaround is to make the original
+ * symbol weak and the alias symbol will automatically become weak too. */
+/* Hint: use `nm -m obj.o` to check the symbols weak/strong on Mac */
+#define __weak_reference(sym,alias) \
```

Comment:
Does this work on Mac OS X with clang? This could replace the .asm stuff with compiler internal bits, which seems like a good change to me:

#define __weak_reference_attribute(sym) __attribute__((__weak__, __alias__(#sym)))
#define __weak_reference(sym,aliassym)	\
	extern __typeof (sym) aliassym __weak_reference_attribute(sym)

Reply:
The problem is that the alias attribute is not supported on Mac. I did try a number of other combinations of inline asm and function attributes but nothing that worked except for this variant. There is evidence online that there are other Mac users that do the straightforward `. weak_reference` asm pseudo op, but either my assembler is broken, or the other users did not check that it actually generates weak references. I will try a few different versions of the tool chain tomorrow to rule out a buggy `llvm-as` build.

Reply:
Ok, so the compiler doesn't work correctly; working around as you have seems like the best option. One thing that confused me with the patch was that it seemed to change a bit more than just adding an os-x specific path for __weak_reference. Could just be diff generating something that's harder to read. I may switch the non-Mac paths to using the compiler attributes though as that should reduce the variants required to support other systems. And, maybe someday the Mac compiler will gain alias attribute support and we could use that.

Reply:
One reason for the confusing diff was that the __weak_reference macro definitions were moved out of the other conditionals, so that the other macros kept the old behavior. I couldn't find any uses of __warn_reference so I figured that it would not be critical to getting this running on native macos.

Reply:
ah, that makes sense. Using 30 year old code ends up accumulating a lot of ancient crufty fixes; I wonder how many of those weird code paths can be merged into a single GCC path that uses attribute now. In any case, once you're happy, and CI passes, I'll review it more closely and then merge if it looks like it leaves things unchanged for the existing paths.

Reply:
So I tested a few different versions of Clang on my Mac:
- Apple clang-1200.0.32.2 (Xcode compiler)
- LLVM Clang 10.0.0 (binary from releases.llvm.org)
- LLVM Clang 11.0.0 (binary from releases.llvm.org)

All of them exhibit this behavior where I need to set the original symbol as weak in order to get the exported alias weak as well.

I tried these variations and some other permutations:
```
#define __weak_reference(sym,alias) \
	__asm__(".weak_reference _" #alias); \
	__asm__(".globl _" #alias); \
	__asm__(".set _" #alias ", _" #sym)
```
```
#define __weak_reference(sym,alias) \
	__asm__(".weak_definition _" #alias); \
	__asm__(".globl _" #alias); \
	__asm__(".set _" #alias ", _" #sym)
```
```
#define __weak_reference(sym,alias) \
	__asm__(".globl _" #alias); \
	__asm__(".set _" #alias ", _" #sym); \
	__asm__(".weak_definition _" #alias)
```

All of them result in this symbol list:
```
nm -m newlib/libc/ssp/libssp.a.p/stack_protector.c.o
0000000000000036 (__TEXT,__text) external ___stack_chk_fail
0000000000000036 (__TEXT,__text) external ___stack_chk_fail_weak
0000000000000748 (__DATA,__common) external ___stack_chk_guard
0000000000000000 (__TEXT,__text) external ___stack_chk_init
                 (undefined) external __exit
                 (undefined) weak external _getentropy
                 (undefined) external _puts
                 (undefined) external _raise
```
and the build error:
```
FAILED: test/stack-smash
clang  -o test/stack-smash test/stack-smash.p/stack-smash.c.o -Wl,-dead_strip_dylibs -Wl,-headerpad_max_install_names -Wl,-undefined,error newlib/libm/libm.a newlib/libc/libc.a -include picolibc.h -fno-builtin -fstack-protector-all -fstack-protector-strong -DTESTS_ENABLE_STACK_PROTECTOR
duplicate symbol '___stack_chk_fail' in:
    test/stack-smash.p/stack-smash.c.o
    newlib/libc/libc.a(stack_protector.c.o)
ld: 1 duplicate symbol for architecture x86_64
clang-11: error: linker command failed with exit code 1 (use -v to see invocation)
```

Using `.weak_definition` on the original symbol (as the proposed patch) results in this symbol list:
```
nm -m newlib/libc/ssp/libssp.a.p/stack_protector.c.o
0000000000000036 (__TEXT,__text) weak external ___stack_chk_fail
0000000000000036 (__TEXT,__text) weak external ___stack_chk_fail_weak
0000000000000748 (__DATA,__common) external ___stack_chk_guard
0000000000000000 (__TEXT,__text) external ___stack_chk_init
                 (undefined) external __exit
                 (undefined) weak external _getentropy
                 (undefined) external _puts
                 (undefined) external _raise
```
removing the `.globl` op on the new alias results in the alias still being strong, but not exported so I get an error about undefined symbol in other places instead (where the test does not override the default weak symbol).

One weird thing I noticed is that even putting the `.weak_defintion` pseudo op after the `.set` still results in both symbols changing into weak when it is applied to the original symbol, but does nothing when applied to the new alias.